File: BT/Module2/Day2/Vertical/1st.java
package Module2.Day2.Vertical;



import java.util.*;



class TreeNode {
    int val;
    TreeNode left, right;
    TreeNode(int x) { val = x; }
}

// Custom Pair class for BFS
class Pair<U, V> {
    public final U key;
    public final V value;
    
    public Pair(U key, V value) {
        this.key = key;
        this.value = value;
    }

    public U getKey() { return key; }
    public V getValue() { return value; }
}

class Solution {
    public List<List<Integer>> verticalTraversal(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if (root == null) return ans;

        Queue<Pair<TreeNode, Pair<Integer, Integer>>> q = new LinkedList<>();
        q.offer(new Pair<>(root, new Pair<>(0, 0))); // (node, (row, col))

        TreeMap<Integer, TreeMap<Integer, PriorityQueue<Integer>>> map = new TreeMap<>();

        while (!q.isEmpty()) {
            Pair<TreeNode, Pair<Integer, Integer>> front = q.poll();
            TreeNode node = front.getKey();
            int row = front.getValue().getKey();
            int col = front.getValue().getValue();

            map.putIfAbsent(col, new TreeMap<>());
            map.get(col).putIfAbsent(row, new PriorityQueue<>());
            map.get(col).get(row).offer(node.val);

            if (node.left != null) q.offer(new Pair<>(node.left, new Pair<>(row + 1, col - 1)));
            if (node.right != null) q.offer(new Pair<>(node.right, new Pair<>(row + 1, col + 1)));
        }

        for (TreeMap<Integer, PriorityQueue<Integer>> colMap : map.values()) {
            List<Integer> vertical = new ArrayList<>();
            for (PriorityQueue<Integer> nodes : colMap.values()) {
                while (!nodes.isEmpty()) {
                    vertical.add(nodes.poll());
                }
            }
            ans.add(vertical);
        }
        return ans;
    }

    // Main method to test the BFS approach
    public static void main(String[] args) {
        Solution sol = new Solution();

        // Creating a sample binary tree
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);
        root.right.left = new TreeNode(6);
        root.right.right = new TreeNode(7);

        // Testing BFS verticalTraversal
        System.out.println("Vertical Order (BFS): " + sol.verticalTraversal(root));
    }
}

























// package Vertical;

// import java.util.*;

// class TreeNode {
//     int val;
//     TreeNode left, right;
//     TreeNode(int x) { val = x; }
// }

// class Solution {
//     public List<List<Integer>> verticalTraversal(TreeNode root) {
//         List<List<Integer>> ans = new ArrayList<>();
//         if (root == null) return ans;

//         Queue<AbstractMap.SimpleEntry<TreeNode, AbstractMap.SimpleEntry<Integer, Integer>>> q = new LinkedList<>();
//         q.offer(new AbstractMap.SimpleEntry<>(root, new AbstractMap.SimpleEntry<>(0, 0))); // (node, (row, col))

//         TreeMap<Integer, TreeMap<Integer, PriorityQueue<Integer>>> map = new TreeMap<>();

//         while (!q.isEmpty()) {
//             AbstractMap.SimpleEntry<TreeNode, AbstractMap.SimpleEntry<Integer, Integer>> front = q.poll();
//             TreeNode node = front.getKey();
//             int row = front.getValue().getKey();
//             int col = front.getValue().getValue();

//             map.putIfAbsent(col, new TreeMap<>());
//             map.get(col).putIfAbsent(row, new PriorityQueue<>());
//             map.get(col).get(row).offer(node.val);

//             if (node.left != null) 
//                 q.offer(new AbstractMap.SimpleEntry<>(node.left, new AbstractMap.SimpleEntry<>(row + 1, col - 1)));
//             if (node.right != null) 
//                 q.offer(new AbstractMap.SimpleEntry<>(node.right, new AbstractMap.SimpleEntry<>(row + 1, col + 1)));
//         }

//         for (TreeMap<Integer, PriorityQueue<Integer>> colMap : map.values()) {
//             List<Integer> vertical = new ArrayList<>();
//             for (PriorityQueue<Integer> nodes : colMap.values()) {
//                 while (!nodes.isEmpty()) {
//                     vertical.add(nodes.poll());
//                 }
//             }
//             ans.add(vertical);
//         }
//         return ans;
//     }

//     public static void main(String[] args) {
//         Solution sol = new Solution();

//         // Creating a sample binary tree
//         TreeNode root = new TreeNode(1);
//         root.left = new TreeNode(2);
//         root.right = new TreeNode(3);
//         root.left.left = new TreeNode(4);
//         root.left.right = new TreeNode(5);
//         root.right.left = new TreeNode(6);
//         root.right.right = new TreeNode(7);

//         // Testing BFS verticalTraversal
//         System.out.println("Vertical Order (BFS): " + sol.verticalTraversal(root));
//     }
// }

File: BT/Module2/Day4/root_leaf/1st.cpp
// class Solution {
//     public:
    
//       void solve(Node* root,vector<vector<int>>&ans,vector<int>arr)
//       {
//           if(!root) return;
//           arr.push_back(root->data);
//           if(root->left==NULL&&root->right==NULL){
//               ans.push_back(arr);
//               return;
//           }
          
//           solve(root->left,ans,arr);
//           solve(root->right,ans,arr);
//       }  
//       vector<vector<int>> Paths(Node* root) {
//           vector<vector<int>>ans;
          
          
//           vector<int>arr;
//           solve(root,ans,arr);
//           return ans;
//       }
//   };







#include <iostream>
#include <vector>
#include <stack>

using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;
    Node(int x) : data(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    vector<vector<int>> Paths(Node* root) {
        if (!root) return {};
        
        vector<vector<int>> ans;
        stack<pair<Node*, vector<int>>> st;
        st.push({root, {}});

        while (!st.empty()) {
            auto front  = st.top();
            Node* node  = front.first;
            auto path  = front.second;
            st.pop();

            path.push_back(node->data);

            if (!node->left && !node->right) {
                ans.push_back(path);
                continue;
            }

            if (node->right) st.push({node->right, path});
            if (node->left) st.push({node->left, path});
        }

        return ans;
    }
};

int main() {
    Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    root->right->left = new Node(6);
    root->right->right = new Node(7);

    Solution sol;
    vector<vector<int>> paths = sol.Paths(root);

    for (auto& path : paths) {
        for (int val : path) cout << val << " ";
        cout << endl;
    }

    return 0;
}











#include <iostream>
#include <vector>
#include <queue>

using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;
    Node(int x) : data(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    vector<vector<int>> Paths(Node* root) {
        if (!root) return {};

        vector<vector<int>> ans;
        queue<pair<Node*, vector<int>>> q;
        q.push({root, {}});

        while (!q.empty()) {
            auto front  = q.front();
            Node* node  = front.first;
            auto path  = front.second;
            q.pop();

            path.push_back(node->data);

            if (!node->left && !node->right) {
                ans.push_back(path);
                continue;
            }

            if (node->left) q.push({node->left, path});
            if (node->right) q.push({node->right, path});
        }

        return ans;
    }
};

int main() {
    Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    root->right->left = new Node(6);
    root->right->right = new Node(7);

    Solution sol;
    vector<vector<int>> paths = sol.Paths(root);

    for (auto& path : paths) {
        for (int val : path) cout << val << " ";
        cout << endl;
    }

    return 0;
}
